### ISSUE TITLE
Does OR-Tools support relative MIP gap configuration?

### ISSUE THREAD

**Issue by jianzhe-luo** [google/or-tools#603]
Hi,

I was trying to solve a large MIP problem using CBC_MIXED_INTEGER_PROGRAMMING. The problem needs to be solved within a couple of seconds while optimality is the secondary objective. So I wonder if there is a way to configure the tolerance level of the optimality gap in MIP solver? Found this old thread https://groups.google.com/forum/#!topic/or-tools-discuss/iWz16p6q680 but it did not solve my problem. Thanks!

**Comment by bacalfa** [comment]
I have the same question for CPLEX. For example, from pages 53/54 [here](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.6.3/ilog.odms.studio.help/pdf/paramcplex.pdf):

![image](https://user-images.githubusercontent.com/1297074/36990464-c6fd7940-2072-11e8-859e-80418095d488.png)

So I tried:

```
solver = pywraplp.Solver("My Model", pywraplp.Solver.CPLEX_MIXED_INTEGER_PROGRAMMING)
solver.SetSolverSpecificParametersAsString("CPXPARAM_MIP_Tolerances_MIPGap  0")
```

but got warnings:

```
WARNING: Logging before InitGoogleLogging() is written to STDERR
W0305 12:43:51.001435 39588 linear_solver.cc:1478] Error in SetSolverSpecificParametersAsString() for solver type: CPLEX_MIXED_INTEGER_PROGRAMMING
W0305 12:43:51.042434 39588 linear_solver.cc:1478] Error in SetSolverSpecificParametersAsString() for solver type: CPLEX_MIXED_INTEGER_PROGRAMMING
```

Similar thing happens for CBC with parameter `AllowablePercentageGap`.

**Comment by danmarinescu** [comment]
I ran into the same issue, so I went ahead and had a look at the implementation of `SetSolverSpecificParametersAsString`. As I understand it from https://github.com/google/or-tools/blob/master/ortools/linear_solver/linear_solver.cc#L1516 the parameters in the string are passed to `SetSolverSpecificParametersAsString` are simply written to a file. Multiple calls to `SetSolverSpecificParametersAsString` simply overwrite the content of that file. Three lines below at https://github.com/google/or-tools/blob/master/ortools/linear_solver/linear_solver.cc#L1519 that file is read with `ReadParameterFile`. This function is implemented differently for each solver. Here is the implementation for CPLEX: https://github.com/google/or-tools/blob/master/ortools/linear_solver/cplex_interface.cc#L1087. As you can see there it calls `CPXXreadcopyparam` which is documented [here](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.6.2/ilog.odms.cplex.help/refcallablelibrary/cpxapi/readcopyparam.html). So the CPXXreadcopyparam` function expects as input a file in the PRM format described [here](https://www.ibm.com/support/knowledgecenter/SSSA5P_12.5.0/ilog.odms.cplex.help/CPLEX/FileFormats/topics/PRM.html). Here is an example of a PRM file:

```
CPLEX Parameter File Version 12.3.0.0
CPX_PARAM_EPPER                  3.45000000000000e-06
CPX_PARAM_OBJULIM                1.23456789012345e+05
CPX_PARAM_PERIND                 1
CPX_PARAM_SCRIND                 1
CPX_PARAM_WORKDIR                "tmp"
```

In conclusion, to be able to pass specific parameters as string to CPLEX one must pass a string to `SetSolverSpecificParametersAsString` which represents the entire content of a valid PRM file. In your case, you need to call it like this:
```
solver.SetSolverSpecificParametersAsString("CPLEX Parameter File Version 12.3.0.0 \n CPXPARAM_MIP_Tolerances_MIPGap 0")
```

Note that based on your CPLEX version you might have a different value for `CPLEX Parameter File Version`

**Comment by lperron** [comment]
It is a parameter of the solver.
